### ISSUE TITLE
Potential bug in GRO code for Windows?

### ISSUE THREAD

**Issue by thomaseizinger** [quinn-rs/quinn#2041]
We are seeing an issue on a Windows machine where `quinn-udp` gives us packets larger than the configured MTU (2540 and 3810). Here is how we split the packets: https://github.com/firezone/firezone/blob/5437c3e2df001c8c491749c7bff0c063ccdf39e1/rust/socket-factory/src/lib.rs#L210-L255

This code has been working on Linux for several months and we have also not had reports that it doesn't work on Windows. Recently however a customer filed a bug report and looking through the logs, I discovered that we receive packets (i.e. `DatagramIn` in the above snippet) that are much larger than expected.

Going through `quinn-udp`'s source code, I developed two theories and would be interested in hearing your thoughts on it:

1. `stride` is being read incorrectly (endianess-issue?) and instead of being smaller than `len`, it is bigger than `len`, meaning we never end up splitting the packets despite it being coalesced. https://github.com/quinn-rs/quinn/blob/d23e4e494f7446e21184bf58acd17a861ae73bba/quinn-udp/src/windows.rs#L241
2. The buffer for control messages is too small, thus Windows fails to fit the `UDP_COALESCED_INFO` message into it and can't tell us that the buffer is coalesced even though it is.

Unfortunately, I've not been able to reproduce this on my own hardware :(

**Comment by stormshield-damiend** [comment]
Does the windows run on intel CPU or on ARM CPU ?

**Comment by mxinden** [comment]
We see a similar issue on Windows on ARM, see summary in [last comment on this thread](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c39).

**Comment by thomaseizinger** [comment]
> Does the windows run on intel CPU or on ARM CPU ?

We see this issue on an x64_86 CPU.

**Comment by thomaseizinger** [comment]
> > The buffer for control messages is too small, thus Windows fails to fit the `UDP_COALESCED_INFO` message into it and can't tell us that the buffer is coalesced even though it is.
> 
> Interesting. I will try reproducing the issue with a larger buffer. For what it is worth, here is [MSQUIC's buffer length](https://github.com/microsoft/msquic/blob/e3ffc71a375d85b1216348b2ffa1e13653213e26/src/platform/datapath_winuser.c#L149-L158).

I thought about this a bit more and I am no longer sure that is a viable possibility unless we receive a lot of other control messages that we don't match on.

**Comment by stormshield-damiend** [comment]
Could the receiver interface be a dual stack (IPv4 and IPv6) one ? I had no chance to test that case when writing the code, maybe we receive ipv4 + ipv6 control messages ?

**Comment by thomaseizinger** [comment]
> We see a similar issue on Windows on ARM, see summary in [last comment on this thread](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c39).

Interesting, according to https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c43, the `UDP_COALESCED_INFO` never gets received. We only get `IP_PKTINFO` and `IP_ECN`.

**Comment by thomaseizinger** [comment]
In https://learn.microsoft.com/en-us/windows-hardware/drivers/network/udp-rsc-offload, it is mentioned that this only works on Windows 11, version 24H2 and later. I'll have to investigate what version our customer is using.

**Comment by thomaseizinger** [comment]
> Could the receiver interface be a dual stack (IPv4 and IPv6) one ? I had no chance to test that case when writing the code, maybe we receive ipv4 + ipv6 control messages ?

I can see in our logs that this customer did have dual-stack but I don't think it matters here because we way we do dual-stack is by having two sockets, one for IPv4 and one for IPv6.

**Comment by stormshield-damiend** [comment]
This code work fine on our side on Windows 10 and 11.

**Comment by stormshield-damiend** [comment]
I do not see any special case related to this in msquic C code https://github.com/microsoft/msquic/blob/main/src/platform/datapath_winuser.c#L3216

**Comment by thomaseizinger** [comment]
@mxinden Do you know by any chance if the affected user is using a dock? We have reports that this only happens when the user is connected via a Dell dock WD19S. It works fine over WiFi and via a separate Ethernet dongle.

**Comment by stormshield-damiend** [comment]
@mxinden could your colleague print the stride value (at L210) before the loop ? it is init with the length of the received packet to handle the case were received packet is not coalesced ?

**Comment by saschanaz** [comment]
The colleague here 👋

I'm not using any dock but only the builtin WiFi, and the stride value is 1280.

**Comment by mxinden** [comment]
> and the stride value is 1280.

Just to avoid any confusion, the stride value is `1280` AND the datagram length value is `1280`. This is with the unit test I provided in [this commit](https://github.com/mxinden/quinn/commit/eb2be2d9842fe2b1039229f104958eba186af9db), see also the logs [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c43). That said, I have the suspicion that the unit test simply doesn't force datagram coalescing in the kernel,  [see this bugzilla comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c43).

Looking at pcaps with fosstodon.org that @saschanaz provided, we see >1_500k UDP datagrams over an internet path. >1_500k over an internet path is **very** unlikely. My suspicion is, that the Windows kernel correctly coalesces the UDP datagrams (< 1_500k) into one >1_500k datagram, but then either does not provide the stride length, or we fail to read the stride length.

**Comment by stormshield-damiend** [comment]
I think you wanted to link this commit https://github.com/mxinden/quinn/commit/7ce09c3f9523920682163c35fca17b4f3a4ed2f3

**Comment by stormshield-damiend** [comment]
We have a difference with msquic, they set the max segment size to "u16:MAX - 8", we set it to "u16:MAX", maybe you can try with this (i doubt it will change anything but who knows) https://github.com/microsoft/msquic/blob/e3ffc71a375d85b1216348b2ffa1e13653213e26/src/platform/datapath_winuser.c#L1496

**Comment by mxinden** [comment]
@saschanaz when you have time, could you (1) apply the following git diff to your local mozilla-central checkout, (2) build in debug mode, and (3) try whether you can reproduce the issues with fosstodon.org. If you can reproduce the issue, can you provide logs and a pcap?

(I wrote the above against 79808865c472858740158ce5cbbf4e0b002cf848, but any commit from today on `bookmarks/central` should work.)

Also see [past Bugzilla comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c5) for log level.

``` diff
diff --git a/third_party/rust/quinn-udp/src/windows.rs b/third_party/rust/quinn-udp/src/windows.rs
index 886f455117ceb..054e267d97921 100644
--- a/third_party/rust/quinn-udp/src/windows.rs
+++ b/third_party/rust/quinn-udp/src/windows.rs
@@ -114,7 +114,7 @@ impl UdpSocketState {
             // u32 per
             // https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-udp-socket-options.
             // Choice of 2^16 - 1 inspired by msquic.
-            u16::MAX as u32,
+            u16::MAX as u32 - 8,
         );
 
         let now = Instant::now();
@@ -211,6 +211,7 @@ impl UdpSocketState {
 
         let cmsg_iter = unsafe { cmsg::Iter::new(&wsa_msg) };
         for cmsg in cmsg_iter {
+            error!("cmsg_iter {} {}", cmsg.cmsg_level, cmsg.cmsg_type);
             const UDP_COALESCED_INFO: i32 = WinSock::UDP_COALESCED_INFO as i32;
             // [header (len)][data][padding(len + sizeof(data))] -> [header][data][padding]
             match (cmsg.cmsg_level, cmsg.cmsg_type) {
@@ -251,6 +252,7 @@ impl UdpSocketState {
             ecn: EcnCodepoint::from_bits(ecn_bits as u8),
             dst_ip,
         };
+        error!("{:?}", meta[0]);
         Ok(1)
     }
 
@@ -398,7 +400,7 @@ fn set_socket_option(
 
 pub(crate) const BATCH_SIZE: usize = 1;
 // Enough to store max(IP_PKTINFO + IP_ECN, IPV6_PKTINFO + IPV6_ECN) + max(UDP_SEND_MSG_SIZE, UDP_COALESCED_INFO) bytes (header + data) and some extra margin
-const CMSG_LEN: usize = 128;
+const CMSG_LEN: usize = 512;
 const OPTION_ON: u32 = 1;
 
 // FIXME this could use [`std::sync::OnceLock`] once the MSRV is bumped to 1.70 and upper
```

@stormshield-damiend the diff above contains your suggestion, namely to use `u16::MAX - 8`. 

@thomaseizinger this includes your suspicion that `CMSG_LEN` might be too small.

Thank you for the help everyone!

**Comment by saschanaz** [comment]
```
 2:34.94 error: the listed checksum of `D:\gecko\third_party\rust\quinn-udp\src/windows.rs` has changed:
 2:34.95 expected: ab1928d18bed62162a0f2c96158d808d7a2962045ab47c9efa0ecf60e2a2c060
 2:34.95 actual:   3a894014bd45e81f7617efe6e3d1c1c2e07a75da68646e7cc8dc447cc984c522
 2:34.95 directory sources are not intended to be edited, if modifications are required then it is recommended that `[patch]` is used with a forked copy of the source
```

I guess I need a fork to make that work?

**Comment by thomaseizinger** [comment]
More theories:

- Do we have a bug in `cmsg::Iter` and fail to decode the last control message? Perhaps printing the raw bytes sitting in the cmesg buffer might be helpful.
- > We have a difference with msquic, they set the max segment size to "u16:MAX - 8", we set it to "u16:MAX"

  The microsoft docs recommend to use [`WSASetUdpRecvMaxCoalesced`](https://learn.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-wsasetudprecvmaxcoalescedsize) instead of the socket option. Maybe that is worth a try?

- > This code work fine on our side on Windows 10 and 11.
  
  And you see coalesced packets on a regular basis?

**Comment by stormshield-damiend** [comment]
> More theories:
> 
>     * Do we have a bug in `cmsg::Iter` and fail to decode the last control message? Perhaps printing the raw bytes sitting in the cmesg buffer might be helpful.
> 
>     * > We have a difference with msquic, they set the max segment size to "u16:MAX - 8", we set it to "u16:MAX"
>       
>       
>       The microsoft docs recommend to use [`WSASetUdpRecvMaxCoalesced`](https://learn.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-wsasetudprecvmaxcoalescedsize) instead of the socket option. Maybe that is worth a try?
> 

msquic does it with a standard setsockeopt call

>     * > This code work fine on our side on Windows 10 and 11.
>       
>       
>       And you see coalesced packets on a regular basis?

Yes we do, but we did not tested in Windows ARM only Intel.

**Comment by mxinden** [comment]
> Sent! It still showed a lot of dropped packets with decryption failure though.

Thank you @saschanaz for the thorough work!

For others, with the above [diff](https://github.com/quinn-rs/quinn/issues/2041#issuecomment-2476683112) the issue still persists. The pcap shows UDP datagrams with up to 20_000 bytes length over an internet link. As expected, the contained QUIC packet(s) fail to decrypt, given that multiple QUIC packets with short headers are disguised as a single large QUIC packet.

**Comment by mxinden** [comment]
> * > This code work fine on our side on Windows 10 and 11.
>       
>       
>       And you see coalesced packets on a regular basis?

For what it is worth, Firefox Nightly does URO on Windows. While we don't see many coalesced receives, we do see some, e.g. in the 99th percentile 8 or more.

![image](https://github.com/user-attachments/assets/d8543628-38a4-4ec1-9356-b16a44f50e40)


https://glam.telemetry.mozilla.org/fog/probe/networking_http_3_udp_datagram_segments_received/explore?aggType=avg&os=Windows&visiblePercentiles=%5B99.9%2C99%2C95%2C75%2C50%2C25%2C5%5D

**Comment by djc** [comment]
@nibanks would you happen to have any guidance/hot tips for this stuff?

**Comment by nibanks** [comment]
Is there a specific problem you need help with? I don't have time to go through the whole thread here.

**Comment by mxinden** [comment]
@nibanks summary:

- reproducible on Windows on ARM (Surface Pro 11 (Qualcomm Snapdragon X))
- also reported on Windows x86_64 
- on a UDP socket [with `UDP_RECV_MAX_COALESCED_SIZE` set](https://github.com/quinn-rs/quinn/blob/d23e4e494f7446e21184bf58acd17a861ae73bba/quinn-udp/src/windows.rs#L113C22-L118)
- [`WSARECVMSG` with GRO](https://github.com/quinn-rs/quinn/blob/d23e4e494f7446e21184bf58acd17a861ae73bba/quinn-udp/src/windows.rs#L186-L192) returns > 1_500 UDP datagrams on Internet path [without `UDP_COALESCED_INFO` cmsg](https://github.com/quinn-rs/quinn/blob/d23e4e494f7446e21184bf58acd17a861ae73bba/quinn-udp/src/windows.rs#L238-L242)

Note that this is the same issue as the one discussed on Slack / e-mail last week, see mail _UDP Receive Segment Coalescing Offload (URO) on Windows on ARM_.

**Comment by saschanaz** [comment]
Per my understanding the URO thing requires 24H2, right? We probably need to make sure that when requesting repro.

**Comment by thomaseizinger** [comment]
@mxinden Have you checked what the return value of https://github.com/quinn-rs/quinn/blob/f8b8c5032e0db9d7dbc7c3452f09c7d1e2a4295d/quinn-udp/src/windows.rs#L110-L118 is on the problematic machine?

Maybe this actually fails but leaves the driver in a kind of limbo state and we should attempt to unset this on failure?

Also, what do people think of adding a safe-guard mechanism like we do for GSO? If the read packet exceeds the interface MTU (I saw y'all have been collaborating on https://github.com/mozilla/mtu), we attempt to unset the socket option.

**Comment by Ralith** [comment]
> If the read packet exceeds the interface MTU we attempt to unset the socket option.

It sounds like that would make sense as a last-ditch mitigation, though it'd be nice to understand the root cause better, especially since it sounds like other software might be dodging this issue somehow. Finding the right interface MTU to refer to might also be complex in practice, since it may change over time and/or depending on address.

**Comment by mxinden** [comment]
> @mxinden Have you checked what the return value of
> 
> https://github.com/quinn-rs/quinn/blob/f8b8c5032e0db9d7dbc7c3452f09c7d1e2a4295d/quinn-udp/src/windows.rs#L110-L118

Yes. @saschanaz tested this, see [this Bugzilla comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c38). `set_socket_option` did not return an error.

> > If the read packet exceeds the interface MTU we attempt to unset the socket option.
> 
> It sounds like that would make sense as a last-ditch mitigation, though it'd be nice to understand the root cause better, especially since it sounds like other software might be dodging this issue somehow. Finding the right interface MTU to refer to might also be complex in practice, since it may change over time and/or depending on address.

Agreed. To add to the above, even if a UDP datagram is smaller than the interface MTU, it might still contain two smaller coalesced UDP datagrams.

> though it'd be nice to understand the root cause better,

For what it is worth, I ordered a Surface Pro 11 (Qualcomm Snapdragon X) yesterday, hoping to be able to reproduce the issue myself. More to come!

**Comment by saschanaz** [comment]
> > @mxinden Have you checked what the return value of
> > https://github.com/quinn-rs/quinn/blob/f8b8c5032e0db9d7dbc7c3452f09c7d1e2a4295d/quinn-udp/src/windows.rs#L110-L118
> 
> Yes. @saschanaz tested this, see [this Bugzilla comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c38). `set_socket_option` did not return an error.

I don't think that was about the return value? But I did test it separately and it didn't fail.

**Comment by mxinden** [comment]
I can now consistently reproduce the bug. It is tied to WSL. Installing WSL makes it occur. Disabling WSL "fixes" the issue.

Here is a _smallish_ reproducer.

On a Windows on ARM device:

1. Install Windows subsystem for Linux in a terminal via `wsl --install`. Note that we are not actually going to use it.
2. Restart Windows.
3. Checkout `https://github.com/mxinden/h3.git` branch [`reproduce-quinn-2041`](https://github.com/mxinden/h3/tree/reproduce-quinn-2041) e.g. using Powershell, no need to use WSL.
4. Run `cargo run --example client -- https://fosstodon.org`.
5. You will see the following panic `Got un-coalesced UDP datagram > 1_500 bytes.`

Once reproduced, you can "fix" the bug by disabling WSL:
1. Control Panel -> Windows features -> turn off "Virtual Machine Platform".
2. Restart Windows.
4. Run `cargo run --example client -- https://fosstodon.org`.
5. See no panic.

@stormshield-damiend do you have an idea why it might fail with WSL installed, but not without?

Note that I am using https://fosstodon.org here which is using Fastly, which consistently sends packet trains (2 or more UDP datagrams back-to-back), which are then coalesced by the client OS. This is not tied to https://fosstodon.org or Fastly.

**Comment by thomaseizinger** [comment]
Very interesting! Does that also reproduce in a VM? Or only on real hardware?

**Comment by mxinden** [comment]
> Does that also reproduce in a VM?

We were not able to reproduce on the following machines, despite WSL being installed:
- Windows on ARM VM running on an ARM MacBook
- Windows X86_64 running natively on a x86_64 CPU

**Comment by stormshield-damiend** [comment]
I have no idea why this fails when WSL is installed

**Comment by mxinden** [comment]
For the record, I am proposing to disable URO on Windows on ARM, at least for now, see https://github.com/quinn-rs/quinn/pull/2071.

While I can consistently reproduce the bug (see https://github.com/quinn-rs/quinn/issues/2041#issuecomment-2495419003), I don't know how to fix the root cause itself.

**Comment by larseggert** [comment]
@nibanks  ^^^

**Comment by thomaseizinger** [comment]
> For the record, I am proposing to disable URO on Windows on ARM, at least for now, see #2071.
> 
> While I can consistently reproduce the bug (see [#2041 (comment)](https://github.com/quinn-rs/quinn/issues/2041#issuecomment-2495419003)), I don't know how to fix the root cause itself.

I'd like to propose an additional (automated?) way to disable URO to workaround this. We got another telemetry alert where we received packets on a machine that are bigger than the MTU. I'd like to either have an API where I can disable URO in that case or have quinn-udp disable it for me automatically.

Whilst this check might not detect it in every case as there might be UDP packets read in a batch where they are in total less than the MTU, **chances are** there will be at least a few packets that maxed out the MTU and are read in a single batch by the driver. In this particular bug report, there were 15 of those.

Thoughts?

**Comment by thomaseizinger** [comment]
> it might delay connection establishment

Slightly OT but why do you think that? Does the initial handshake not fit into a UDP packet? Why would the initial handshake send multiple packets that are not filled to the MTU?

There will still be a delay of course because the coalesced packets needs to be retransmitted. I would expect the actual handshake to succeed though. At least that was what we saw in the logs: Connection established fine and then "breaks" (once the first packet train arrives).

**Comment by mxinden** [comment]
> > it might delay connection establishment
> 
> Slightly OT but why do you think that? Does the initial handshake not fit into a UDP packet? Why would the initial handshake send multiple packets that are not filled to the MTU?

You are right. One would not send multiple coalesced packets each smaller than the MTU.

---

For the record only. In the buggy setups discussed above, Windows does not set the `UDP_COALESCED_INFO` header on a coalesced read. Thus I wondered whether non-buggy setups would always set the `UDP_COALESCED_INFO` header when URO is enabled. In the case of a non-coalesced read `UDP_COALESCED_INFO` would simply equal the overall UDP datagram size. That would allow us to differentiate the two setups (buggy vs. non-buggy) and thus disable URO on buggy setups. I tested this. Unfortunately the answer is no, a non-buggy setup only sets `UDP_COALESCED_INFO` on coalesced reads.

**Comment by thomaseizinger** [comment]
> > > it might delay connection establishment
> > 
> > 
> > Slightly OT but why do you think that? Does the initial handshake not fit into a UDP packet? Why would the initial handshake send multiple packets that are not filled to the MTU?
> 
> You are right. One would not send multiple coalesced packets each smaller than the MTU.

Should we devise a strategy then where we configure an MTU (or hardcode one?) and disable GRO if we receive a packet larger than it without `UDP_COALESCED_INFO` being set?

**Comment by mxinden** [comment]
I am not sure how that would work.

Say the local network interface MTU is 1500 bytes. The network path of the connection has a bottleneck of 1280 bytes somewhere, which the local endpoint isn't aware of. Say we assume a reasonable local MTU of 1350 bytes. On an incoming datagram of 1350 bytes we could then not differentiate a valid single 1350 bytes datagram from a buggy 1280 bytes + 170 bytes coalesced datagram.  

This gets even worse once considering jumbo frames on LANs.

**Comment by thomaseizinger** [comment]
> You are right. One would not send multiple coalesced packets each smaller than the MTU.

I do think this particular case can happen. For example, multiple STUN requests to the same peer could indeed be coalesced into one packet.

In general though, I think it is quite rare and there would _eventually_ be a coalesced packet > MTU on which we would detect the buggy driver.

I'll need to do some more perf testing to see if it is worth it. Most of our customers are on Windows so it would be a nice boost. I'll see if I can implement something in Firezone to test this strategy.

**Comment by thomaseizinger** [comment]
> On an incoming datagram of 1350 bytes we could then not differentiate a valid single 1350 bytes datagram from a buggy 1280 bytes + 170 bytes coalesced datagram.

Yeah it is definitely not bulletproof and depends a lot on the traffic pattern. I guess the safest would be to perform some kind of load-test to check whether the driver can handle GRO correctly.

**Comment by thomaseizinger** [comment]
It has been a while since I dealt with this. Were we able to reproduce this with a unit-test on the buggy hardware?

**Comment by mxinden** [comment]
@thomaseizinger the smallest reproducer I am aware of is https://github.com/quinn-rs/quinn/issues/2041#issuecomment-2495419003.

Maybe it is possible to reproduce via GitHub's `windows-11-arm` now. I haven't tried. Related, we are running into other issues with these runners, see https://github.com/mozilla/neqo/pull/2591.

**Comment by thomaseizinger** [comment]
> [@thomaseizinger](https://github.com/thomaseizinger) the smallest reproducer I am aware of is [#2041 (comment)](https://github.com/quinn-rs/quinn/issues/2041#issuecomment-2495419003).

So on this hardware, the unit tests of `quinn-udp` pass just fine? Because should should use GRO if I am not mistaken.

The reason I am asking is because if we can create self-contained test, then we could ship such a function with `quinn-udp` and perform a self-test when a new socket is created. Like:

- bind two random ports
- send a packet train from one socket to another
- observed if we can correctly handle it

If yes, we know the hardware is safe to use with GRO, if no, we can disable it.

That would be a much more reliable way of conditionally enabling it because we don't need to "wait" for a traffic pattern that triggers the detection as we can just generate the traffic ourselves and therefore know, how big the packets should be that we are reading.

**Comment by ehoogeveen-medweb** [comment]
FWIW, in the [linked Firefox issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1916558) I found that receive segment coalescing needed to be specifically available and enabled in the relevant network drivers:
* Realtek: `Recv Segment Coalescing` needs to be present and enabled[^1]
* Intel: `RSCv4` and/or `RSCv6` needs to be present and enabled[^2] (presumably they're for IPv4 and IPv6 respectively)

And of course URO needs to be available in Windows - I only started having issues after updating to Windows 11 24H2.
If you run `netsh int udp show global` it should show the URO state as `Receive Offload State` (though this is only the Windows feature, the drivers still have to agree).

[^1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c59
[^2]: https://bugzilla.mozilla.org/show_bug.cgi?id=1916558#c62

**Comment by mxinden** [comment]
> So on this hardware, the unit tests of `quinn-udp` pass just fine?

Yes, back then we were not able to reproduce the bug on the buggy devices using the `quinn-udp` unit tests. The unit tests do use URO. I assume the bugs didn't surface, as they are using localhost networking, i.e. don't go through the (presumably buggy) network card (drivers).

**Comment by BarabasGitHub** [comment]
Hello, I'm having a peculiar issue and this is the only place I've found anything that looks similar.
We have a situation where only some end users end up receiving concatenated udp datagrams from a recv call. For them this happens consistently, though far from all the time. I've been scratching my head a a lot and went through our code, adding debug logging etc. I've never seen anything like this happen before. I know udp can duplicate, reorder and drop packets, but it should never give me two from a recv call. 
This coalescing option is the only potential candidate I've been able to identify, however we don't enable or disable it explicitly. When I use getsockopt with UDP_RECV_MAX_COALESCED_SIZE it tells me it's 0. Any chance it's still possible to trigger this bug?
This is on windows x64.

**Comment by mxinden** [comment]
@BarabasGitHub please open a new issue. I don't think this is related. Please include the `quinn-udp` version you are using.

**Comment by BarabasGitHub** [comment]
@mxinden sorry, I'm not using quinn, but this is the only similar problem I found that resembles what goes wrong in our application (using plain udp). So I was hoping someone here knows more about this kind of bug where you receive multiple udp datagrams in one recv.
